File: studentassistant/src/main/java/ru/erdenian/studentassistant/activity/ScheduleActivity.java
package ru.erdenian.studentassistant.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.view.PagerTabStrip;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.Toast;

import com.codetroopers.betterpickers.calendardatepicker.CalendarDatePickerDialogFragment;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSortedSet;

import org.joda.time.LocalDate;
import org.joda.time.LocalTime;

import ru.erdenian.studentassistant.R;
import ru.erdenian.studentassistant.adapter.SchedulePagerAdapter;
import ru.erdenian.studentassistant.schedule.Homework;
import ru.erdenian.studentassistant.schedule.Lesson;
import ru.erdenian.studentassistant.schedule.OnScheduleUpdateListener;
import ru.erdenian.studentassistant.schedule.ScheduleManager;
import ru.erdenian.studentassistant.schedule.Semester;
import ru.erdenian.studentassistant.ulils.UiUtils;

/**
 * Todo: описание класса.
 *
 * @author Ilya Solovyev
 * @version 0.0.0
 * @since 0.0.0
 */
public class ScheduleActivity extends AppCompatActivity implements
        AdapterView.OnItemSelectedListener,
        CalendarDatePickerDialogFragment.OnDateSetListener,
        View.OnClickListener,
        OnScheduleUpdateListener {

    private static final String CURRENT_PAGE = "current_page";

    private int savedPage = -1;

    private Semester selectedSemester;

    private DrawerLayout drawer;
    private Spinner spSemesters;
    private LinearLayout llAddButtons;
    private ViewPager viewPager;
    private SchedulePagerAdapter pagerAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_schedule);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar_with_spinner);
        setSupportActionBar(toolbar);

        drawer = UiUtils.initializeDrawerAndNavigationView(this, R.id.activity_schedule_drawer,
                toolbar, getResources());

        spSemesters = (Spinner) findViewById(R.id.toolbar_with_spinner_spinner);
        spSemesters.setOnItemSelectedListener(this);

        llAddButtons = (LinearLayout) findViewById(R.id.content_schedule_add_buttons);
        Button btnGetScheduleFromServer = (Button) findViewById(R.id.content_schedule_get_schedule_from_server);
        btnGetScheduleFromServer.setOnClickListener(this);
        Button btnAddSchedule = (Button) findViewById(R.id.content_schedule_add_schedule);
        btnAddSchedule.setOnClickListener(this);

        viewPager = (ViewPager) findViewById(R.id.view_pager);
        PagerTabStrip pagerTabStrip = (PagerTabStrip) findViewById(R.id.view_pager_pager_tab_strip);
        pagerTabStrip.setTextColor(ContextCompat.getColor(this, R.color.colorPrimary));
        pagerTabStrip.setTabIndicatorColorResource(R.color.colorPrimary);
    }

    @Override
    protected void onStart() {
        super.onStart();
        ScheduleManager.setOnScheduleUpdateListener(this);
        onScheduleUpdate();
    }

    @Override
    public void onScheduleUpdate() {
        getSupportActionBar().setDisplayShowTitleEnabled(ScheduleManager.getSemesters().size() <= 1);
        spSemesters.setVisibility((ScheduleManager.getSemesters().size() > 1) ? View.VISIBLE : View.GONE);

        viewPager.setVisibility((ScheduleManager.getSemesters().size() > 0) ? View.VISIBLE : View.GONE);
        llAddButtons.setVisibility((ScheduleManager.getSemesters().size() == 0) ? View.VISIBLE : View.GONE);

        if ((pagerAdapter != null) && (selectedSemester.getId() == ScheduleManager.getSelectedSemester().getId())) {
            savedPage = viewPager.getCurrentItem();
        }
        selectedSemester = ScheduleManager.getSelectedSemester();

        if (ScheduleManager.getSemesters().size() > 1) {
            ArrayAdapter<String> adapter = new ArrayAdapter<>(this, R.layout.spinner_item_semesters, ScheduleManager.getSemestersNames());
            adapter.setDropDownViewResource(R.layout.spinner_dropdown_item_semesters);
            spSemesters.setAdapter(adapter);
            spSemesters.setSelection(ScheduleManager.getSelectedSemesterIndex());
        } else if (ScheduleManager.getSemesters().size() == 1) {
            getSupportActionBar().setTitle(selectedSemester.getName());
            onItemSelected(null, null, 0, 0);
        } else {
            getSupportActionBar().setTitle(R.string.title_activity_schedule);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle bundle) {
        bundle.putInt(CURRENT_PAGE, viewPager.getCurrentItem());
        super.onSaveInstanceState(bundle);
    }

    @Override
    protected void onRestoreInstanceState(Bundle bundle) {
        savedPage = bundle.getInt(CURRENT_PAGE, -1);
        super.onRestoreInstanceState(bundle);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_schedule, menu);
        menu.findItem(R.id.menu_schedule_calendar).setVisible(!ScheduleManager.getSemesters().isEmpty());
        UiUtils.colorMenu(this, menu);
        return true;
    }

    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
        ScheduleManager.setSelectedSemesterIndex(i);
        pagerAdapter = new SchedulePagerAdapter(getSupportFragmentManager(), selectedSemester);
        viewPager.setAdapter(pagerAdapter);
        viewPager.setCurrentItem((savedPage != -1) ? savedPage : pagerAdapter.getPosition(LocalDate.now()), false);
        savedPage = -1;
    }

    @Override
    public void onNothingSelected(AdapterView<?> adapterView) {
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_schedule_calendar:
                UiUtils.showDatePicker(selectedSemester.getFirstDay(), selectedSemester.getLastDay(), LocalDate.now(),
                        getSupportFragmentManager(), this);
                break;
            case R.id.menu_schedule_edit_schedule:
                startActivity(new Intent(this, SemestersEditorActivity.class));
                break;
            default:
                Log.wtf(this.getClass().getName(), "Неизвестный id: " + item.getItemId());
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDateSet(CalendarDatePickerDialogFragment dialog, int year, int monthOfYear, int dayOfMonth) {
        viewPager.setCurrentItem(pagerAdapter.getPosition(new LocalDate(year, monthOfYear + 1, dayOfMonth)));
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.content_schedule_get_schedule_from_server:
                Toast.makeText(this, R.string.activity_schedule_get_schedule_from_server_button, Toast.LENGTH_SHORT).show();
                break;
            case R.id.content_schedule_add_schedule:

                ImmutableSortedSet<Lesson> lessons = ImmutableSortedSet.of(
                        new Lesson("Конструирование ПО", "Лабораторная работа",
                                ImmutableSortedSet.of("Федоров Алексей Роальдович", "Федоров Петр Алексеевич"), ImmutableSortedSet.of("4212а"),
                                new LocalTime(14, 20), new LocalTime(15, 50),
                                Lesson.RepeatType.BY_WEEKDAY, 5, ImmutableList.of(false, true), null),
                        new Lesson("Конструирование ПО", "Лабораторная работа",
                                ImmutableSortedSet.of("Федоров Алексей Роальдович"), ImmutableSortedSet.of("4212а"),
                                new LocalTime(18, 10), new LocalTime(19, 40),
                                Lesson.RepeatType.BY_WEEKDAY, 5, ImmutableList.of(false, true), null));

                ImmutableSortedSet<Semester> semesters = ImmutableSortedSet.of(
                        new Semester("Семестр 5", new LocalDate(2016, 9, 1), new LocalDate(2016, 12, 31),
                                lessons, ImmutableSortedSet.<Homework>of()));

                ScheduleManager.setSemesters(semesters);

                break;
            default:
                Log.wtf(this.getClass().getName(), "Неизвестный id: " + view.getId());
                break;
        }
    }

    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }
}

File: studentassistant/src/main/kotlin/ru/erdenian/studentassistant/activity/ScheduleActivity.kt
package ru.erdenian.studentassistant.activity

import android.os.Bundle
import android.support.v4.view.GravityCompat
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import kotlinx.android.synthetic.main.activity_schedule.*
import kotlinx.android.synthetic.main.toolbar_with_spinner.*
import kotlinx.android.synthetic.main.view_pager.*
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast
import org.joda.time.LocalDate
import ru.erdenian.studentassistant.R
import ru.erdenian.studentassistant.adapter.SchedulePagerAdapter
import ru.erdenian.studentassistant.extensions.getCompatColor
import ru.erdenian.studentassistant.extensions.initializeDrawerAndNavigationView
import ru.erdenian.studentassistant.extensions.setColor
import ru.erdenian.studentassistant.extensions.showDatePicker
import ru.erdenian.studentassistant.localdata.ScheduleManager

class ScheduleActivity : AppCompatActivity(),
    AdapterView.OnItemSelectedListener,
    ScheduleManager.OnScheduleUpdateListener {

  companion object {
    private const val CURRENT_PAGE = "current_page"
    private const val SELECTED_SEMESTER_ID = "selected_semester_id"

    private const val BUTTONS_INDEX = 0
    private const val SCHEDULE_INDEX = 1
  }

  private var savedPage: Int? = null
  private var selectedSemesterId: Long? = null

  private var pagerAdapter: SchedulePagerAdapter? = null

  private val actionBar by lazy { supportActionBar ?: throw IllegalStateException("supportActionBar == null") }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_schedule)

    setSupportActionBar(toolbar_with_spinner)
    initializeDrawerAndNavigationView(toolbar_with_spinner)

    toolbar_with_spinner_spinner.onItemSelectedListener = this

    content_schedule_get_schedule_from_server.setOnClickListener {
      toast(R.string.content_schedule_get_schedule_from_server_button)
    }
    content_schedule_add_schedule.setOnClickListener {
      startActivity<SemesterEditorActivity>()
    }

    view_pager_pager_tab_strip.setTextColor(getCompatColor(R.color.colorPrimary))
    view_pager_pager_tab_strip.setTabIndicatorColorResource(R.color.colorPrimary)
  }

  override fun onStart() {
    super.onStart()
    ScheduleManager.addOnScheduleUpdateListener(this)
    onScheduleUpdate()
  }

  override fun onScheduleUpdate() {
    if (ScheduleManager.semesters.size <= 1) {
      actionBar.setDisplayShowTitleEnabled(true)
      toolbar_with_spinner_spinner.visibility = View.GONE
    } else {
      actionBar.setDisplayShowTitleEnabled(false)
      toolbar_with_spinner_spinner.visibility = View.VISIBLE
    }

    if (ScheduleManager.semesters.isNotEmpty()) {
      schedule_flipper.displayedChild = SCHEDULE_INDEX
    } else {
      schedule_flipper.displayedChild = BUTTONS_INDEX
    }

    if ((pagerAdapter != null) && (selectedSemesterId == ScheduleManager.selectedSemesterId)) {
      savedPage = view_pager.currentItem
    }
    selectedSemesterId = ScheduleManager.selectedSemesterId

    when {
      ScheduleManager.semesters.size > 1 -> {
        val adapter = ArrayAdapter(this, R.layout.spinner_item_semesters, ScheduleManager.semestersNames)
        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item_semesters)
        toolbar_with_spinner_spinner.adapter = adapter
        toolbar_with_spinner_spinner.setSelection(ScheduleManager.selectedSemesterIndex)
      }
      ScheduleManager.semesters.size == 1 -> {
        actionBar.title = ScheduleManager.selectedSemester.name
        onItemSelected(null, null, 0, -1L)
      }
      else -> {
        actionBar.setTitle(R.string.title_activity_schedule)
        pagerAdapter = null
        view_pager.adapter = null
      }
    }

    invalidateOptionsMenu()
    initializeDrawerAndNavigationView(toolbar_with_spinner)
  }

  override fun onSaveInstanceState(outState: Bundle) {
    outState.putInt(CURRENT_PAGE, view_pager.currentItem)
    outState.putLong(SELECTED_SEMESTER_ID, selectedSemesterId ?: -1L)
    super.onSaveInstanceState(outState)
  }

  override fun onRestoreInstanceState(savedInstanceState: Bundle) {
    savedPage = savedInstanceState.getInt(CURRENT_PAGE, -1).takeIf { it != -1 }
    selectedSemesterId = savedInstanceState.getLong(SELECTED_SEMESTER_ID, -1L).takeIf { it != -1L }
    ScheduleManager.selectedSemesterId = selectedSemesterId
    super.onRestoreInstanceState(savedInstanceState)
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.menu_schedule, menu)

    val isSemestersEmpty = ScheduleManager.semesters.isEmpty()
    menu.findItem(R.id.menu_schedule_calendar).isVisible = !isSemestersEmpty
    menu.findItem(R.id.menu_schedule_edit_schedule).isVisible = !isSemestersEmpty

    menu.setColor(getCompatColor(R.color.action_bar_icons_color))
    return true
  }

  override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
    selectedSemesterId = ScheduleManager.semesters.asList()[position].id
    ScheduleManager.selectedSemesterId = selectedSemesterId

    pagerAdapter = SchedulePagerAdapter(supportFragmentManager, ScheduleManager.selectedSemester, false)
    view_pager.adapter = pagerAdapter
    view_pager.setCurrentItem(savedPage ?: pagerAdapter!!.getPosition(LocalDate.now()), false)
    savedPage = null
  }

  override fun onNothingSelected(parent: AdapterView<*>) = Unit

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.menu_schedule_calendar -> ScheduleManager.selectedSemester.let {
        showDatePicker(pagerAdapter?.getDate(view_pager.currentItem), it.firstDay, it.lastDay) { newDate ->
          pagerAdapter?.run { view_pager.currentItem = getPosition(newDate) }
        }
      }
      R.id.menu_schedule_add_schedule -> startActivity<SemesterEditorActivity>()
      R.id.menu_schedule_edit_schedule -> startActivity<LessonsEditorActivity>(
          SEMESTER_ID to ScheduleManager.selectedSemesterId
      )
      else -> throw IllegalArgumentException("Неизвестный id: ${item.itemId}")
    }
    return super.onOptionsItemSelected(item)
  }

  override fun onBackPressed() {
    if (drawer_layout.isDrawerOpen(GravityCompat.START)) drawer_layout.closeDrawer(GravityCompat.START)
    else super.onBackPressed()
  }
}
